src/components/impagination.js

```diff
@@ -0,0 +1,149 @@
+      observe: this.observe
+    });
+
+    // initial state
+    this.state = {
+      datasetState: this.dataset.state
+    };
+  }
+
+  // set the initial read offset of our dataset
+  componentWillMount() {
+    let { readOffset } = this.props;
+    this.dataset.setReadOffset(readOffset);
+  }
+
+  // when we recieve an update, loop over the pages of collections so
+  // we can resolve or reject any pending requests
+  componentWillReceiveProps(nextProps) {
+    let { totalPages, collection } = nextProps;
+
+    // new, zero length collections are likely entirely new
+    let isNewCollection = !collection.length &&
+      collection.isLoading !== this.props.collection.isLoading;
+
+    // we need to reset the dataset
+    if (isNewCollection) {
+      this.promises = {};
+      this.dataset.reset();
+    }
+
+    // if there is a new total page count, update our dataset state
+    if (totalPages !== this.dataset.state.stats.totalPages) {
+      this.dataset.state.stats.totalPages = totalPages;
+    }
+
+    // ensure that we keep all of our promises
+    for (let page of Object.keys(this.promises)) {
+      let promise = this.promises[page];
+      let { request, records } = collection.getPage(page);
+
+      if (promise) {
+        if (request.isResolved) {
+          promise.resolve(records);
+          delete this.promises[page];
+        } else if (request.isRejected) {
+          promise.reject(request.errors);
+          delete this.promises[page];
+        }
+      }
+    }
+  }
+
+  componentDidUpdate() {
+    let { readOffset } = this.props;
+
+    // if there is a new read offset, tell our dataset
+    if (readOffset !== this.dataset.state.readOffset) {
+      this.dataset.setReadOffset(readOffset);
+    }
+  }
+
+  /**
+   * Used by impagination's dataset to resolve records
+   * @param {Number} pageOffset - zero based page offset
+   * @param {Number} pageSize - requested page size
+   * @returns {Promise} resolves immediately or during the
+   * componentWillReceiveProps hook later
+   */
+  fetch = (pageOffset, pageSize) => {
+    // page starts with 1
+    let page = pageOffset + 1;
+
+    return new Promise((resolve, reject) => {
+      let { request, records } = this.props.collection.getPage(page);
+
+      // if the collection has already been resolved, immediately
+      // resolve this promise
+      if (request.isResolved) {
+        resolve(records);
+
+        // if the collection has already been rejected, immediately
+        // reject this promise
+      } else if (request.isRejected) {
+        reject(request.errors);
+
+        // the collection is not resolved or rejected already, make
+        // the request now and handle resolution later on in the
+        // componentWillReceiveProps hook
+      } else {
+        this.props.fetch(page, pageSize);
+
+        // cache this promise's callbacks (and status) so it can be
+        // resolved or rejected later on
+        this.promises[page] = { resolve, reject };
```

Comment:
I can't seem to find it now, but I thought I recalled you mentioning something like `dataset.resolvePage(page, records)` and `dataset.rejectPage(page, error)` or something along those lines. Which would remove the need for our promises. In the case where we go with and API like this, would `fetch` then just return undefined?

Reply:
Yeah, sounds like we need to really think this out. Let's leave it as is, and schedule a retrospective once this gets merged so that we can suggest improvements to impagination.